File: 2022/haskell/Day02.hs
{-# LANGUAGE GADTs #-}
{-# LANGUAGE LambdaCase #-}

module Day02 (problem1, problem2) where

import Control.Category ((>>>))
import Data.Functor ((<&>))

data Move = Rock | Paper | Scissors

data Outcome = Win | Lose | Draw

data Problem a where
  Problem1 :: Problem Move
  Problem2 :: Problem Outcome

char2Move :: Char -> Move
char2Move = \case
  'A' -> Rock
  'B' -> Paper
  'C' -> Scissors
  'X' -> Rock
  'Y' -> Paper
  'Z' -> Scissors
  _ -> error "Invalid char"

char2Outcome :: Char -> Outcome
char2Outcome = \case
  'X' -> Lose
  'Y' -> Draw
  'Z' -> Win
  _ -> error "Invalid char"

move2Score :: Move -> Integer
move2Score = \case
  Rock -> 1
  Paper -> 2
  Scissors -> 3

outcome2Score :: Outcome -> Integer
outcome2Score = \case
  Win -> 6
  Draw -> 3
  Lose -> 0

readInput :: FilePath -> Problem a -> IO [(Move, a)]
readInput file problem = do
  readFile file
    <&> ( lines
            >>> fmap words
            >>> fmap
              ( \case
                  [[o], [p]] ->
                    ( char2Move o,
                      case problem of
                        Problem1 -> char2Move p
                        Problem2 -> char2Outcome p
                    )
                  _ -> error "bad input"
              )
        )

problem1 :: FilePath -> IO Integer
problem1 file = do
  readInput file Problem1 <&> sum . fmap score
  where
    play :: (Move, Move) -> Outcome
    play (Rock, Paper) = Win
    play (Scissors, Rock) = Win
    play (Paper, Scissors) = Win
    play (Paper, Rock) = Lose
    play (Rock, Scissors) = Lose
    play (Scissors, Paper) = Lose
    play _ = Draw

    score :: (Move, Move) -> Integer
    score moves@(_, you) =
      move2Score you + outcome2Score (play moves)

problem2 :: FilePath -> IO Integer
problem2 file = do
  readInput file Problem2 <&> sum . fmap score
  where
    play :: (Move, Outcome) -> Move
    play (Rock, Win) = Paper
    play (Scissors, Win) = Rock
    play (Paper, Win) = Scissors
    play (Paper, Lose) = Rock
    play (Rock, Lose) = Scissors
    play (Scissors, Lose) = Paper
    play (move, Draw) = move

    score :: (Move, Outcome) -> Integer
    score theRound@(_, outcome) =
      move2Score (play theRound) + outcome2Score outcome

File: 2022/rust/day02.rs
use crate::utils::throw;
use std::fs;
use std::path;

struct Round {
    opponent: Move,
    player: Move,
    outcome: Outcome,
}

#[derive(PartialEq)]
enum Move {
    Rock,
    Paper,
    Scissors,
}

enum Outcome {
    Win,
    Lose,
    Draw,
}

fn char_to_move(ch: &str) -> Move {
    match ch {
        "A" => Move::Rock,
        "B" => Move::Paper,
        "C" => Move::Scissors,
        "X" => Move::Rock,
        "Y" => Move::Paper,
        "Z" => Move::Scissors,
        _ => throw("bad input"),
    }
}

fn char_to_outcome(ch: &str) -> Outcome {
    match ch {
        "X" => Outcome::Lose,
        "Y" => Outcome::Draw,
        "Z" => Outcome::Win,
        _ => throw("bad input"),
    }
}

fn read_input(file: &path::Path) -> Vec<Round> {
    let contents = fs::read_to_string(file).expect("should be able to read file");
    let mut rounds = Vec::new();
    for line in contents.lines() {
        let split: Vec<&str> = line.split(" ").collect();
        if let &[opponent, player] = &split[..2] {
            rounds.push(Round {
                opponent: char_to_move(opponent),
                player: char_to_move(player),
                outcome: char_to_outcome(player),
            })
        } else {
            throw("bad input")
        }
    }
    rounds
}

fn score_outcome(outcome: &Outcome) -> u32 {
    match outcome {
        Outcome::Win => 6,
        Outcome::Draw => 3,
        Outcome::Lose => 0,
    }
}

fn score_move(m: &Move) -> u32 {
    match m {
        Move::Rock => 1,
        Move::Paper => 2,
        Move::Scissors => 3,
    }
}

pub fn problem1(file: &path::Path) -> u32 {
    let rounds = read_input(file);
    let mut result = 0;
    for round in rounds {
        let Round {
            opponent, player, ..
        } = round;
        result += score_move(&player);
        match (&opponent, &player) {
            (Move::Rock, Move::Paper) => result += score_outcome(&Outcome::Win),
            (Move::Paper, Move::Scissors) => result += score_outcome(&Outcome::Win),
            (Move::Scissors, Move::Rock) => result += score_outcome(&Outcome::Win),
            _ => {
                if opponent == player {
                    result += score_outcome(&Outcome::Draw);
                }
            }
        }
    }
    result
}

pub fn problem2(file: &path::Path) -> u32 {
    let rounds = read_input(file);
    let mut result = 0;
    for round in rounds {
        let Round {
            opponent, outcome, ..
        } = round;
        result += score_outcome(&outcome);
        match (&opponent, &outcome) {
            (Move::Rock, Outcome::Win) => result += score_move(&Move::Paper),
            (Move::Paper, Outcome::Win) => result += score_move(&Move::Scissors),
            (Move::Scissors, Outcome::Win) => result += score_move(&Move::Rock),
            (Move::Rock, Outcome::Lose) => result += score_move(&Move::Scissors),
            (Move::Paper, Outcome::Lose) => result += score_move(&Move::Rock),
            (Move::Scissors, Outcome::Lose) => result += score_move(&Move::Paper),
            (_, Outcome::Draw) => result += score_move(&opponent),
        }
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn problem1_sample_input() {
        let result = problem1(path::Path::new("problems/day02/input_sample.txt"));
        assert_eq!(result, 15);
    }

    #[test]
    fn problem1_real_input() {
        let result = problem1(path::Path::new("problems/day02/input.txt"));
        assert_eq!(result, 11150);
    }

    #[test]
    fn problem2_sample_input() {
        let result = problem2(path::Path::new("problems/day02/input_sample.txt"));
        assert_eq!(result, 12);
    }

    #[test]
    fn problem2_real_input() {
        let result = problem2(path::Path::new("problems/day02/input.txt"));
        assert_eq!(result, 8295);
    }
}
